src/analysis/souffle/examples/multimic/multimic-preferences.authlogic

```diff
@@ -0,0 +1,335 @@
+// When "UserX" enters the room, a runtime on the user's
+// mobile device sends a packet to the microphone which contains the following
+// facts serialized (and with relevant signatures):
+// "MobileDeviceRuntime" says {
+//     hasTag("MicIn.audio", "UserX_userSpeech").
+// }
+// - Also passes along the facts by "UserX" (which are signed with UserX's key).
+//
+
+// The end users all delegate to a policy principal that decides what can be
+// done with their data. Delegation to this policy principal happens when they 
+// buy the device or decide to create an account within this privacy framework.
+// It can be thought of as a mechanized EULA. The users then express their 
+// preferences about how their data can be handled to the policy principal,
+// but these preferences are not always met.
```

Comment:
I recognize the value of this EULA pattern; the expression of the multimic policy in this file is much more auth-logic-idiomatic and rich than examples that do not use it. It does feel a little like this is manufacturing a centralized authority in a decentralized system; you could ask whether an EULA that is tied to buying the device offers users meaningful control over their privacy if there are only two real phone OS choices in the world.

I think you could potentially avoid the EULA-based approach by using the ownership tags that @bgogul introduced. This would involve the following steps:

1. Analyze the mic's subgraph in isolation, and from the combination of the mic's subgraph and UserA's `stay_on` policy, push to the edges of the subgraph the requirement that any owner that flows into the system allows for `userSpeech` to be downgraded to `textTranscript`.

2. Analyze the smartphone's graph in isolation and propagate ownership tags to the edges of the (presumably small) smartphone graph.

3. Upon connection, check the ownership tag against the mic's requirement.

I think that might serve our purpose without needing the EULA. That said, I see the value of the EULA pattern; perhaps we could develop both of these approaches as separate examples. I recognize additionally that the EULA pattern works right now, whereas the more static-analysis heavy approach requires more mechanism, so perhaps this should be the canonical example for a bit.

Reply:
> It does feel a little like this is manufacturing a centralized authority in a decentralized system; you could ask whether an EULA that is tied to buying the device offers users meaningful control over their privacy if there are only two real phone OS choices in the world.

I agree with this, actually. The whole point of the way policy checking works with a DIFC-ish taint analysis, authorization logic, and the "may/will" check is that absolutely everything that happens to data owned by some party (e.g. a user) can only be done with this data owner's consent. The PRD directly contrasts the way this policy checking normally works, so it is a pretty difficult case to handle. By default, this policy system would normally require every user to agree to everything that might happen with their data.

The mechanical part of this pattern just works by having all the data owners consent to having their preferences overridden under some situations (by doing a delegation). The "EULA" explanation was just my way of connecting this to something tied in reality.